### ISSUE TITLE
Advice request on optimization of deep learning network

### ISSUE THREAD

**Issue by feevos** [SheffieldML/GPyOpt#178]
Dear all, 

I am addressing the community to ask for some advice on the results of some trial runs I've performed. In particular I would please like some advice if the algorithm is exploring parameter space sensibly or there is something I can improve. 

I am experimenting with BOpt for optimizing the hyperparameters of a deep learning architecture. Specifically, I have a deep neural network for semantic segmentation and I am using the adam optimizer. This has 5 parameters (which I am optimizing), namely weight decay, learning rate, beta1, beta2 and a smoothing criterion epsilon ```params = {wd, learning_rate, beta1, beta2, epsilon}```. 
My prior range for these is: 
```
[(1.0e-8, 1.0e-1), (1.0e-8, 1.0e-2), (0.2, 0.999), (0.2, 0.999), (1.0e-9, 1.0e-5)]
```
For the evaluation of a single set of proposed values vector ```params```, I am calculating a 2D vector, that consists of the weighted mean loss, and a weighted mean numerical derivative. I am using as evaluation criterion the angle (in 2D) plane of the normalized vector as my GPyOpt optimization function (the closer to -90deg <==> higher diff(loss), smaller loss, the better). Anyway, in tests this looks good. 

Now, perhaps cruicially, in every iteration of the NN, I start from the same weights, but I feed **randomly different data points**. I don't train for a whole epoch, but I use ~ 100 batch updates. That is, with a batch size of 256 (so the NN sees 25600 images befor GPyOpt decides for the value of the objective function. These images differ in general from run to run. So there is noise in the optimization function, that is, **if the algorithm tries to re-evaluate the same proposed points it will end up with a different objective function value.** This is why I am using ```de_duplication=True``` (is this correct?). 

For 5 dimensions I am evaluating in total 5*12+1 objective function evaluations. According to these [guidelines](https://blog.sigopt.com/posts/evaluating-hyperparameter-optimization-strategies) one needs approximately 10-20 objective function evaluations per dimension, so, 1 initial point (sensible proposed Adam values) + 5x12 in total.    The optimization routine is the following

```Python
bounds =  self.config[C.C_BO_BOUNDS]
domain = self.config[C.C_BO_DOMAIN]

X_init = np.array(self.config[C.C_BO_XINIT]) # ADD THIS INTO CONFIG DICT 
Y_init = np.array([[self.evaluate_opt_point(epoch, X_init)]])
print ("Proposed parameters:: {}, angle::{}".format(X_init,Y_init))

X_step = X_init
Y_step = Y_init
for i in range(self.nBOevals):
    bo_step = GPyOpt.methods.BayesianOptimization(f = None,
                                                    model_type='GP_MCMC',
                                                    acquisition_type = 'EI_MCMC',
                                                    domain=domain,
                                                    X = X_step,
                                                    Y = Y_step,
                                                    de_duplication=True, # Avoids evaluating in repeating points 
                                                    num_cores = self.num_cores)

    # Propose new evaluation point. 
    x_next = bo_step.suggest_next_locations()
    y_next = np.array([[self.evaluate_opt_point(epoch, x_next)]])
    print ("Proposed parameters:: {}, angle::{}".format(x_next,y_next))
    # Stack new evaluation points to previous ones. 
    X_step = np.vstack((X_step, x_next))
    Y_step = np.vstack((Y_step, y_next))
```
In the following image you can see the proposed values from the algorithm for each variable seperately. The horizontal axis is the step of the evaluation (in total 61 steps).  In the last (bottom right) plot overplotted with red are the 6 best solutions that GPyOpt finds.  
![image](https://user-images.githubusercontent.com/14290519/38967947-f13a9e9c-43bb-11e8-86af-f05714851105.png)

In this plot you can see paired plots of the variables. 
![image](https://user-images.githubusercontent.com/14290519/38967468-b508d080-43b9-11e8-90a1-a1a1439166eb.png)

The algorithm proposes sensible points in the end. 

My questions: 
1. Do the parameters look like sensibly sampled? 
2. Do I need to add any more options in the GPyOpt.methods.BayesianOptimization object to account for the noisy nature of the objective function? 

Thank you very much for your time. 

In the following are the top 5 algorithm choices for the parameters: 
```
              wd  learning_rate     beta1  beta2       epsilon  BO_angle
1.000000e-08           0.01  0.999000  0.999  1.399033e-06 -0.059110
1.000000e-08           0.01  0.874904  0.999  1.000000e-09 -0.057122
1.000000e-08           0.01  0.883368  0.999  1.000000e-09 -0.044510
1.000000e-08           0.01  0.914978  0.999  4.592317e-06 -0.044353
1.000000e-08           0.01  0.879413  0.999  1.000000e-09 -0.044047
```

In the following are all the values the algorithm proposes: 
```
wd learning_rate beta1 beta2 epsilon BO_angle
1e-05 0.01 0.9 0.99 1e-06 -0.02937530556059675
0.1 1e-08 0.2 0.2 1e-05 -0.004962479761413029
1e-08 0.01 0.9493611947100283 0.999 3.2223661664371207e-07 -0.032553982915340784
0.1 1e-08 0.999 0.999 1e-05 0.0026218944694064938
1e-08 0.01 0.9352470983054882 0.9598819504142639 1e-09 -0.03172746621397456
1e-08 0.01 0.5491775181458098 0.999 1e-09 -0.0278487594401685
1e-08 0.01 0.8491590566012516 0.9637388115406581 1e-09 -0.03183459433669321
1e-08 0.01 0.2 0.999 1e-09 -0.022878517596082764
1e-08 0.01 0.8759086834130733 0.999 1e-09 -0.04273012842061961
1e-08 1e-08 0.8527450866292297 0.999 1e-09 -0.0008195080445428905
1e-08 0.01 0.7989564022892998 0.999 1e-09 -0.03455850258683958
1e-08 0.01 0.6076815958024134 0.999 1e-09 -0.020331801345363092
1e-08 0.01 0.999 0.999 1e-09 -0.02271493684537537
1.0000000001675335e-08 0.01 0.999 0.999 1e-09 -0.03943545286262716
1e-08 0.01 0.999 0.999 1.3990332381897745e-06 -0.05911042820029484
1e-08 0.01 0.999 0.999 2.627521272354336e-07 -0.025325292164489813
1e-08 0.01 0.999 0.999 1.458077365602501e-06 -0.027625857295347545
0.0021196348593543373 0.006078787067208551 0.9558307307574594 0.9916240437481554 4.162632840638215e-06 -0.03877434500428092
1e-08 0.01 0.999 0.999 6.719154224711783e-07 -0.02056931651169786
1e-08 0.01 0.994312952651229 0.999 1e-09 -0.019751728998561403
1e-08 0.01 0.999 0.999 8.785106988769042e-07 -0.026059391587854672
1e-08 0.01 0.999 0.8282723113438608 1e-09 -0.029528393409652746
0.015623157986359678 0.008203452775485834 0.9878808312019562 0.8986076963258278 3.564798859200664e-06 -0.022890718047544792
1e-08 0.01 0.999 0.999 1.5881106018036974e-06 -0.035077863111776846
1e-08 0.01 0.8823398729427729 0.999 9.820779844908284e-06 -0.03943391573946202
1e-08 0.01 0.9531879292088896 0.999 1e-05 -0.03295362581305068
1e-08 0.006799014367319078 0.9547837375489961 0.9838159709919133 4.415550935975974e-06 -0.019078860621918412
1e-08 0.01 0.9164676242693282 0.999 1e-05 -0.035728140113090126
1e-08 0.01 0.8787234417204399 0.999 3.549552963531736e-06 -0.04132741988678974
1e-08 0.01 0.999 0.999 1.3402727107176275e-06 -0.028294302873843914
1e-08 0.007554310961630244 0.999 0.999 1e-05 -0.015281406010867167
1e-08 0.01 0.999 0.999 8.099133984436972e-07 -0.015184243831407955
1e-08 0.01 0.887653379080321 0.999 1e-05 -0.02993847780465798
0.0027843557104722458 0.008951175157847746 0.939527796274148 0.9772684337386524 3.308926702660501e-06 -0.038622710114280465
1e-08 0.01 0.8795037821307669 0.999 1e-09 -0.03142939090124281
1e-08 0.01 0.9085865966443856 0.999 6.780405094858257e-06 -0.0275517773839136
1e-08 0.01 0.999 0.42868986608850346 1e-09 -0.013307969299335991
0.01318108570227709 0.004059171153904448 0.7899503736221047 0.9988313395569952 8.887716252868048e-06 -0.022191607813796057
1e-08 0.01 0.9149782943819909 0.999 4.5923173729128245e-06 -0.04435343201602908
1e-08 0.01 0.918544319388446 0.999 1e-09 -0.01620788398744288
1e-08 0.01 0.9163120755443689 0.999 1e-05 -0.022917953645478183
1e-08 0.01 0.8796605251194699 0.999 9.168168570568724e-07 -0.036676864200267875
0.06158626328762336 0.0064130797330330365 0.9962040352645984 0.9935993920468449 1.3306699921826887e-06 -0.007017628616415346
1e-08 0.01 0.8749040836147044 0.999 1e-09 -0.05712224333188463
1e-08 0.01 0.8761323631850302 0.999 1e-09 -0.03020165215358916
1e-08 0.01 0.7370711889432434 0.999 1e-09 -0.007488162725126238
1e-08 0.01 0.8794134963628134 0.999 1e-09 -0.0440468385264653
1e-08 0.01 0.8833675314829179 0.999 1e-09 -0.044509905110579985
1e-08 0.01 0.999 0.999 2.2502275324056403e-06 -0.02500154052843444
1e-08 0.01 0.8792748866709581 0.999 1e-09 -0.03249761527260943
1e-08 0.01 0.8828327743305465 0.999 9.999999999996516e-10 -0.029994615599962977
1e-08 0.01 0.8830083733855251 0.999 1e-09 -0.030125357973152616
1e-08 0.01 0.8836153233244131 0.999 1e-09 -0.035460650615005467
1e-08 0.01 0.8819231260692841 0.999 1e-09 -0.03857864495934765
1e-08 0.01 0.8493774522432515 0.7941310459298676 1e-09 -0.021743610616178913
1e-08 0.01 0.38379072014609283 0.7289939575806011 1e-05 -0.029368145560459382
1e-08 0.01 0.8854144985329656 0.999 1e-09 -0.03957831668364534
1e-08 0.01 0.8786620426364418 0.999 1e-09 -0.02068587256520039
1e-08 0.01 0.8779733679723688 0.999 2.3217399098712238e-06 -0.03628259484679559
1e-08 0.01 0.8866572430239885 0.999 1e-09 -0.0349270152297596
0.1 0.01 0.4215940473140339 0.999 8.253178938759313e-06 -0.0045618731509679605

```
